File: Python/Algorithm/Algorithm I/733. Flood Fill.py
import collections
from typing import List


class Solution:
    def isValidCell(self, row: int, column: int, ROWS: int, COLUMNS: int) -> bool:
        """
        Check if the given cell coordinates are within the bounds of the grid.

        Args
            row (int): Row index of the cell.
            column (int): Column index of the cell.
            ROWS (int): Total number of rows in the grid.
            COLUMNS (int): Total number of columns in the grid.

        Returns
            bool: True if the cell is valid, False otherwise.

        Time Complexity: o(1) since we do 2 comparisons.
        Space Complexity: o(1) since we do not use any extra space.
        """
        isValidRow = 0 <= row < ROWS
        isValidColumn = 0 <= column < COLUMNS

        return isValidRow and isValidColumn

    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        """
        Perform flood fill on the given image starting from the specified coordinates.

        Args
            image (List[List[int]]): The 2D grid representing the image.
            sr (int): The starting row index for flood fill.
            sc (int): The starting column index for flood fill.
            color (int): The new color to be filled.

        Returns
            List[List[int]]: The modified image after flood fill.

        Time Complexity: o(m * n), where m is the number of rows and n is the number of columns in the image.
        Space Complexity: o(m * n), where m is the number of rows and n is the number of columns in the image.
        """
        ROWS = len(image)
        COLUMNS = len(image[0])
        DIRECTIONS = [(-1, 0), (1, 0), (0, -1), (0, 1)]

        oldColor = image[sr][sc]
        newColor = color

        if oldColor == newColor:
            return image

        queue = collections.deque([[sr, sc]])
        image[sr][sc] = newColor

        while queue and queue[0]:
            row, column = queue.popleft()
            
            for rowStep, columnStep in DIRECTIONS:
                newRow = row + rowStep
                newColumn = column + columnStep

                if not self.isValidCell(newRow, newColumn, ROWS, COLUMNS):
                    continue

                if image[newRow][newColumn] != oldColor:
                    continue

                image[newRow][newColumn] = newColor
                queue.append([newRow, newColumn])

        return image

File: Java/Algorithm/Algorithm I/733. Flood Fill.java
class Solution {

  /**
   * Checks if a cell is a valid cell within the given image.
   *
   * @param image  The image grid to check cell validity within.
   * @param row    The row index of the cell.
   * @param column The column index of the cell.
   * @return True if the cell is valid, otherwise false.
   */
  private boolean isValidCell(int[][] image, int row, int column) {
    int ROWS = image.length;
    int COLUMNS = image[0].length;

    boolean isValidRow = row >= 0 && row < ROWS;
    boolean isValidColumn = column >= 0 && column < COLUMNS;

    return isValidRow && isValidColumn;
  }

  /**
   * Performs depth-first search (DFS) to fill an area of the image with a new color.
   *
   * @param image    The image grid to be filled.
   * @param row      The current row index.
   * @param column   The current column index.
   * @param oldColor The original color to be replaced.
   * @param newColor The new color to fill the area with.
   */
  private void dfs(int[][] image, int row, int column, int oldColor, int newColor) {
    if (isValidCell(image, row, column)) {
      if (image[row][column] == oldColor) {
        image[row][column] = newColor;

        dfs(image, row - 1, column, oldColor, newColor);
        dfs(image, row + 1, column, oldColor, newColor);
        dfs(image, row, column - 1, oldColor, newColor);
        dfs(image, row, column + 1, oldColor, newColor);
      }
    }
  }

  /**
   * Fills the area of the image starting from a specified cell with a new color.
   *
   * @param image The image grid to be filled.
   * @param sr    The row index of the starting cell.
   * @param sc    The column index of the starting cell.
   * @param color The new color to fill the area with.
   * @return The modified image with the filled area.
   */
  public int[][] floodFill(int[][] image, int sr, int sc, int color) {
    int ROWS = image.length;
    int COLUMNS = image[0].length;

    int newColor = color;
    int oldColor = image[sr][sc];

    if (oldColor == newColor) return image;

    dfs(image, sr, sc, oldColor, newColor);

    return image;
  }
}
